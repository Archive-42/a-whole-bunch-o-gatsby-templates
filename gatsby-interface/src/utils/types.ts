// AnyTag is anything that a JSX tag can be.
export type AnyTag =
  | string
  | React.FunctionComponent<never>
  | (new (props: never) => React.Component)

// PropsOf tries to get the expected properties for a given HTML tag name or component.
export type PropsOf<Tag> = Tag extends keyof JSX.IntrinsicElements
  ? JSX.IntrinsicElements[Tag]
  : Tag extends React.ComponentType<infer Props>
  ? Props & JSX.IntrinsicAttributes
  : never
