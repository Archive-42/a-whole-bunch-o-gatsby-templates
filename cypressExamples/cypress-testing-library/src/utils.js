function getFirstElement(jqueryOrElement) {
  if (Cypress.dom.isJquery(jqueryOrElement)) {
    return jqueryOrElement.get(0)
  }
  return jqueryOrElement
}

function getContainer(container) {
  const withinContainer = cy.state('withinSubject')
  if (withinContainer) {
    return getFirstElement(withinContainer)
  }
  return getFirstElement(container)
}

export {getFirstElement, getContainer}

/* globals Cypress, cy */
