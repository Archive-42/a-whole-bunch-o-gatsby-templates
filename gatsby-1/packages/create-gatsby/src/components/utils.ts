import stringLength from "string-length"
// ansi and emoji-safe string length
import wordWrap from "ansi-wordwrap"

const DEFAULT_WIDTH = process.stdout.columns

export const wrap = (text: string, width = DEFAULT_WIDTH): string =>
  wordWrap(text, { width })

export function rule(char = `\u2501`, width = DEFAULT_WIDTH): string {
  return char.repeat(width)
}

export function center(
  text: string,
  padding = ` `,
  width = DEFAULT_WIDTH
): string {
  const pad = padding.repeat(Math.round((width - stringLength(text)) / 2))
  return pad + text
}
