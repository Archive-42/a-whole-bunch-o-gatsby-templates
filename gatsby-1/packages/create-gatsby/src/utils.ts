const ESC = `\u001b`

export const clearLine = (count = 1): Promise<void> =>
  new Promise(resolve => {
    // First move the cursor up one line...
    process.stderr.moveCursor(0, -count, () => {
      // ... then clear that line. This is the ANSI escape sequence for "clear whole line"
      // List of escape sequences: http://ascii-table.com/ansi-escape-sequences.php
      process.stderr.write(`${ESC}[2K`)
      resolve()
    })
  })

// Makes a string safe for using as a folder or npm package name
export const makeNpmSafe = (str: string): string =>
  str
    // Replace camelcase with kebab
    .replace(/([a-z])([A-Z])/g, `$1-$2`)
    .toLowerCase()
    // Replace any number of consecutive illegal characters with a single dash
    .replace(/[^a-z0-9_.]+/g, `-`)
    // Remove trailing dots and dashes
    .replace(/^[_\-.]+|[_\-.]+$/g, ``)
