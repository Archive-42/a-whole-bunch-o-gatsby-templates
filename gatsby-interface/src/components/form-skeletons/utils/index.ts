export function getFinalAriaDescribedBy(
  controlDescribedBy?: string,
  ariaDescribedBy?: string
) {
  return (
    [controlDescribedBy, ariaDescribedBy]
      .filter(describedBy => describedBy)
      .join(` `) || undefined
  )
}

export function getHintId(fieldId: string) {
  return `${fieldId}__hint`
}

export function getErrorId(fieldId: string) {
  return `${fieldId}__error`
}
