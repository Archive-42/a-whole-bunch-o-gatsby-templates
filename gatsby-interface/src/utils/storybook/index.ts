export * from "./knobs"
export * from "./layoutHelpers"
export * from "./withVariationsContainer"
export * from "./propsCombinationStory"
export { default as StoryUtils } from "./StoryUtils"
