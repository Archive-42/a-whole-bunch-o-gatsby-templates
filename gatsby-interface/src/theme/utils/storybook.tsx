/** @jsx jsx */
import { jsx } from "@emotion/core"
import React from "react"
import { Theme } from ".."
import { ThemeProvider } from "../../components/ThemeProvider"
import styled from "@emotion/styled"

export function ThemeDocs({
  theme,
  scale,
}: {
  theme: Theme
  scale?: keyof Theme
}) {
  return (
    <ThemeProvider>
      <div css={{ fontFamily: theme.fonts.monospace }}>
        <h1>{scale || "Theme values"}</h1>
        <p>Token names are sorted alphabetically</p>
        {scale ? (
          <ScaleGrid>{renderScale(scale, theme[scale])}</ScaleGrid>
        ) : (
          Object.entries(theme).map(([scaleName, scaleTokens]) => {
            return (
              <React.Fragment>
                <h2>{scaleName}</h2>
                <ScaleGrid>{renderScale(scaleName, scaleTokens)}</ScaleGrid>
              </React.Fragment>
            )
          })
        )}
      </div>
    </ThemeProvider>
  )
}

const ScaleGrid = styled.div(props => ({
  display: `grid`,
  gridTemplateColumns: `1fr 2fr`,
  rowGap: (props.theme as Theme).space[4],
}))

/**
 * A recursive function that renders a flattened theme scale
 * For example, a scale like this:
 * {
 *   ui: {
 *     background: "#fbfbfb",
 *     border: {
 *       subtle: "#f0f0f2"
 *     }
 *   },
 *   white: "#ffffff",
 * }
 * would be displayed like this:
 *
 * ui.background #fbfbfb
 * ui.border.subtle #f0f0f2
 * white #ffffff
 */
const renderScale = (
  scaleName: string,
  scaleTokens: object,
  tokenPath: string | null = null
): React.ReactNode => {
  return Object.entries(scaleTokens)
    .sort(sortTokens)
    .map(([token, value]) => {
      const currentPath = tokenPath
        ? `${tokenPath}${isNumberKey(token) ? `[${token}]` : `.${token}`}`
        : token

      return typeof value === "object" && value !== null ? (
        renderScale(scaleName, value as Record<string, unknown>, currentPath)
      ) : (
        <TokenDisplay scaleName={scaleName} label={currentPath} value={value} />
      )
    })
}

function Code({ children }: { children: React.ReactNode }) {
  return (
    <span
      css={(theme: Theme) => ({
        color: theme.colors.grey[90],
        backgroundColor: theme.colors.white,
        padding: `${theme.space[1]} ${theme.space[2]}`,
        borderRadius: theme.radii[1],
      })}
    >
      {children}
    </span>
  )
}

function ColorDisplay({ color }: { color: string }) {
  return (
    <span
      css={theme => ({
        display: `inline-block`,
        verticalAlign: `middle`,
        width: theme.fontSizes[1],
        height: theme.fontSizes[1],
        border: `1px solid ${theme.colors.grey[40]}`,
        backgroundColor: color,
        marginRight: theme.space[2],
        borderRadius: theme.radii[1],
      })}
    />
  )
}

function TokenValue({
  scaleName,
  value,
}: {
  scaleName: string
  value: unknown
}) {
  return (
    <React.Fragment>
      {[`colors`, `tones`].includes(scaleName) && typeof value === `string` && (
        <ColorDisplay color={value} />
      )}
      <Code>{typeof value === `string` ? value : JSON.stringify(value)}</Code>
    </React.Fragment>
  )
}

function TokenDisplay({
  scaleName,
  label,
  value,
}: {
  scaleName: string
  label: string
  value: unknown
}) {
  return (
    <React.Fragment>
      <div>
        <Code>{label}</Code>
      </div>
      <div>
        <TokenValue scaleName={scaleName} value={value} />
      </div>
    </React.Fragment>
  )
}

function isNumberKey(value: string) {
  return !isNaN(Number(value))
}

function sortTokens([token1]: [string, unknown], [token2]: [string, unknown]) {
  if (isNumberKey(token1) && isNumberKey(token2)) {
    return Number(token1) - Number(token2)
  }
  if (token1 > token2) {
    return 1
  } else if (token1 < token2) {
    return -1
  }
  return 0
}
