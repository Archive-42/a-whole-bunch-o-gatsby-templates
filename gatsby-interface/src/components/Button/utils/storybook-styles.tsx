/** @jsx jsx */
import { jsx } from "@emotion/core"
import React from "react"

import { action } from "@storybook/addon-actions"
import { MdArrowForward } from "react-icons/md"
import { radios } from "@storybook/addon-knobs"

import { StoryUtils } from "../../../utils/storybook"
import { StoryApi } from "@storybook/addons"
import {
  ButtonSize,
  ButtonTone,
  ButtonVariant,
} from "../../../theme/styles/button"
import { StoryFnReactReturnType } from "@storybook/react/dist/client/preview/types"

function enumToOptions<T extends string>(
  memo: { [k: string]: string },
  value: T
) {
  return { ...memo, [value]: value }
}

const BUTTON_SIZES: ButtonSize[] = [`S`, `M`, `L`, `XL`]
const BUTTON_SIZE_OPTIONS = BUTTON_SIZES.reduce(enumToOptions, {})

const BUTTON_TONES: ButtonTone[] = [`BRAND`, `NEUTRAL`, `SUCCESS`, `DANGER`]
const BUTTON_TONE_OPTIONS = BUTTON_TONES.reduce(enumToOptions, {})

const BUTTON_VARIANTS: ButtonVariant[] = [`PRIMARY`, `SECONDARY`, `GHOST`]
const BUTTON_VARIANT_OPTIONS = BUTTON_VARIANTS.reduce(enumToOptions, {})

type ShowcaseReturn = Parameters<StoryApi<StoryFnReactReturnType>["add"]>

export function showcaseVariants<P>(
  Component: React.ComponentType<P>,
  defaultProps: P
): ShowcaseReturn {
  return [
    `variants`,
    () => {
      const size: ButtonSize = radios(`size`, BUTTON_SIZE_OPTIONS, `L`)
      const tone: ButtonTone = radios(`tone`, BUTTON_TONE_OPTIONS, `BRAND`)

      const renderCase = (variant: ButtonVariant) => (
        <Component
          onClick={action(`Button was clicked`)}
          {...defaultProps}
          variant={variant}
          tone={tone}
          size={size}
          ref={undefined}
        >
          variant&nbsp;<React.Fragment>'{variant}'</React.Fragment>
        </Component>
      )

      return (
        <StoryUtils.Container>
          <StoryUtils.Stack>{BUTTON_VARIANTS.map(renderCase)}</StoryUtils.Stack>
        </StoryUtils.Container>
      )
    },
  ]
}

export function showcaseSizes<P>(
  Component: React.ComponentType<P>,
  defaultProps: P
): ShowcaseReturn {
  return [
    `sizes`,
    () => {
      const variant: ButtonVariant = radios(
        `variant`,
        BUTTON_VARIANT_OPTIONS,
        `PRIMARY`
      )
      const tone: ButtonTone = radios(`tone`, BUTTON_TONE_OPTIONS, `BRAND`)

      const renderCase = (size: ButtonSize) => (
        <Component
          onClick={action(`Button was clicked`)}
          {...defaultProps}
          variant={variant}
          tone={tone}
          size={size}
          ref={undefined}
        >
          size&nbsp;<React.Fragment>'{size}'</React.Fragment>
        </Component>
      )

      return (
        <StoryUtils.Container>
          <StoryUtils.Stack>{BUTTON_SIZES.map(renderCase)}</StoryUtils.Stack>
        </StoryUtils.Container>
      )
    },
  ]
}

export function showcaseTones<P>(
  Component: React.ComponentType<P>,
  defaultProps: P
): ShowcaseReturn {
  return [
    `tones`,
    () => {
      const variant: ButtonVariant = radios(
        `variant`,
        BUTTON_VARIANT_OPTIONS,
        `PRIMARY`
      )
      const size: ButtonSize = radios(`size`, BUTTON_SIZE_OPTIONS, `L`)

      const renderCase = (tone: ButtonTone) => (
        <Component
          onClick={action(`Button was clicked`)}
          {...defaultProps}
          variant={variant}
          tone={tone}
          size={size}
          ref={undefined}
        >
          tone&nbsp;<React.Fragment>'{tone}'</React.Fragment>
        </Component>
      )

      return (
        <StoryUtils.Container>
          <StoryUtils.Stack>{BUTTON_TONES.map(renderCase)}</StoryUtils.Stack>
        </StoryUtils.Container>
      )
    },
  ]
}

export function showcaseCustomStyles<P>(
  Component: React.ComponentType<P>,
  defaultProps: P,
  readme?: string
): ShowcaseReturn {
  return [
    `override/extend styles`,
    () => (
      <StoryUtils.Container>
        <StoryUtils.Stack>
          <Component
            {...defaultProps}
            onClick={action(`Button was clicked`)}
            css={theme => ({
              color: theme.colors.purple[60],
              background: theme.colors.yellow[60],
              borderColor: theme.colors.yellow[60],

              "&:hover:not([disabled])": {
                color: theme.colors.white,
              },
            })}
          >
            Button with custom style
          </Component>
        </StoryUtils.Stack>
      </StoryUtils.Container>
    ),
    readme
      ? {
          readme: {
            sidebar: readme,
          },
        }
      : undefined,
  ]
}

export function showcaseIcons<P>(
  Component: React.ComponentType<P>,
  defaultProps: P,
  readme?: string
): ShowcaseReturn {
  return [
    `with icons`,
    () => (
      <StoryUtils.Container>
        <StoryUtils.Stack>
          <Component
            {...defaultProps}
            onClick={action(`Button was clicked`)}
            rightIcon={<MdArrowForward />}
          >
            On the right
          </Component>
          <Component
            {...defaultProps}
            onClick={action(`Button was clicked`)}
            leftIcon={<MdArrowForward />}
          >
            On the left
          </Component>
        </StoryUtils.Stack>
      </StoryUtils.Container>
    ),
    readme
      ? {
          readme: {
            sidebar: readme,
          },
        }
      : undefined,
  ]
}
