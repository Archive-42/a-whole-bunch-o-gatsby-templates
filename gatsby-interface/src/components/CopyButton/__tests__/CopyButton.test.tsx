import React from "react"
import { fireEvent } from "@testing-library/react"
import { CopyButton } from "../index"
import { renderWithTheme } from "../../../utils/testing"

const render = renderWithTheme

const clipboard = () => {
  let buffer: string | undefined = undefined

  return {
    writeText: async (text: string) =>
      new Promise(resolve => {
        buffer = text
        resolve()
      }),
    readText: async () => new Promise<string>(resolve => resolve(buffer)),
  }
}

let oldClipboard: Clipboard

beforeAll(() => {
  oldClipboard = window.navigator.clipboard

  // @ts-ignore
  window.navigator.clipboard = clipboard()
})

afterAll(() => {
  // @ts-ignore
  window.navigator.clipboard = oldClipboard
})

describe(`<CopyButton>`, () => {
  it(`renders unchanged`, async () => {
    const { container } = render(<CopyButton content="test" />)

    expect(container).toMatchSnapshot()
  })

  it(`allows to configure button label`, async () => {
    const { container } = render(
      <CopyButton
        content="lorem ipsum"
        getButtonLabel={copied => (copied ? `BarCopied` : `FooCopy`)}
      />
    )

    const button = container.querySelector(`button`) as HTMLButtonElement
    expect(button).toHaveTextContent(`FooCopy`)

    fireEvent.click(button)

    await window.navigator.clipboard.readText()

    expect(button).toHaveTextContent(`BarCopied`)
  })

  it(`allows to configure button title`, async () => {
    const { container } = render(
      <CopyButton
        content="lorem ipsum"
        getButtonTitle={copied => (copied ? `BarCopied` : `FooCopy`)}
      />
    )

    const button = container.querySelector(`button`) as HTMLButtonElement
    expect(button).toHaveAttribute(`title`, `FooCopy`)

    fireEvent.click(button)

    await window.navigator.clipboard.readText()

    expect(button).toHaveAttribute(`title`, `BarCopied`)
  })

  it(`copies the passed content to clipboard and updates button label`, async () => {
    const { container } = render(<CopyButton content="lorem ipsum" />)

    const button = container.querySelector(`button`) as HTMLButtonElement
    expect(button).toHaveTextContent(`Copy`)

    fireEvent.click(button)

    let copiedText = ``
    await window.navigator.clipboard
      .readText()
      .then((text: string) => (copiedText = text))

    expect(button).toHaveTextContent(`Copied`)
    expect(copiedText).toBe(`lorem ipsum`)
  })
})
