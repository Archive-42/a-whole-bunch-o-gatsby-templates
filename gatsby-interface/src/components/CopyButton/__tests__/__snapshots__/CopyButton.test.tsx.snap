// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<CopyButton> renders unchanged 1`] = `
<div>
  <button
    class="css-1wkaa75-baseCss"
    title="Copy to clipboard"
    type="button"
  >
    Copy
  </button>
</div>
`;
